lib/galaxy/model/tags.py

```diff
@@ -245,7 +247,18 @@ def _get_tag(self, tag_name):
         return self.sa_session.query(galaxy.model.Tag).filter_by(name=tag_name).first()
 
     def _create_tag_instance(self, tag_name):
-        return galaxy.model.Tag(type=0, name=tag_name)
+        # For good performance caller should first check if there's already an appropriate tag
+        Session = sessionmaker(self.sa_session.bind)
+        tag = galaxy.model.Tag(type=0, name=tag_name)
+        with Session() as separate_session:
+            separate_session.add(tag)
+            try:
+                separate_session.commit()
+                separate_session.flush()
```

Comment:
```suggestion
```
already flushed via `commit()`

Reply:
That was my intuition too, but the unit tests fail without this. I haven't checked out why, but my guess is that the sessionfactory parameters aren't picked up. (and maybe that explains why our similar attempts in `ToolShedRepositoryCache` were doomed to fail ...)

Reply:
My local unit tests are not failing. What is the test that's failing on your setup?

I see that the values of these params `autoflush=False, autocommit=True` are reversed on the created session (i.e., `self.sa_session` vs `separate_session`). I'll try to figure out how to fix this.

Reply:
OK, so getting the session factory params to apply is easy: `Session.configure(autoflush=False, autocommit=True)` is the immediate "fix". Of course, a better way to do it is to have a function somewhere in `model.base` (or similar location) that can be called anytime we need a one-time separate, properly configured session object. Could be a context manager for good measure. 

However, right now that will raise an error (maybe that's the error you were seeing). The problem is the autocommit mode. As per docs:  "When using legacy “autocommit” mode, this method is only valid to call if a transaction is actually in progress, else an error is raised." ([ref](https://docs.sqlalchemy.org/en/14/orm/session_api.html#sqlalchemy.orm.Session.commit))

I think what we have here is sufficient for 21.09. Once we address the autocommit setting (as part of the move to SA 2.0), we can revisit this, etc.